tests/syntax/test_dynamics.py

```diff
@@ -576,12 +594,206 @@ def test_behavior_invariant():
             while True:
                 take self.position.x
                 self.position -= Range(0, 2) @ 0
-        ego = Object at 1 @ 0, with behavior Foo
+        ego = new Object at 1 @ 0, with behavior Foo
     """)
     for i in range(30):
         actions = sampleEgoActions(scenario, maxSteps=3, maxIterations=50)
         assert actions[1] > 0
 
+def test_precondition_rejection():
+    scenario = compileScenic("""
+        box_region = BoxRegion(position=(0,0,0), dimensions=(1,1,1)).getSurfaceRegion()
+
+        # This behavior should work properly since it has no preconditions/invariants
+        behavior BehaviorA():
+            ego.foo += 1
+            wait
+
+        # This behavior should never execute since the precondition will reject,
+        # which should be converted to a PreconditionViolation
+        behavior BehaviorB():
+            precondition: box_region.orientation[self.position].yaw < float('inf')
+
+            ego.bar += 1
+            wait
+
+        behavior MetaBehavior():
+            while True:
+                try:
+                    do BehaviorB()
+                except PreconditionViolation:
+                    do BehaviorA()
+
+        ego = new Object at (0,0,0), with foo 0, with bar 0,
+            with behavior MetaBehavior()
+
+        record final (ego.foo, ego.bar) as test_val
+    """)
+    scene = sampleScene(scenario)
+    result = sampleResultFromScene(scene, maxSteps=20)
+    assert result is not None
+    assert result.records['test_val'] == (20,0)
```

Comment:
Using mutable state and `record` statements for these tests overcomplicates them. You can just have `take 1` in `BehaviorA`, `take 2` in `BehaviorB`, and use `sampleEgoActions` (there's also no need to slow the test down by running the simulation for more than 1 time step). Take a look at `test_choose_1` for example.